rmw_dds_common/src/qos.cpp

```diff
@@ -0,0 +1,313 @@
+  char * reason,
+  size_t reason_size)
+{
+  if (!compatibility) {
+    RMW_SET_ERROR_MSG("compatibility parameter is null");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  if (!reason && reason_size != 0u) {
+    RMW_SET_ERROR_MSG("reason parameter is null, but reason_size parameter is not zero");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  // If there are any "unknown" values, then there is an error
+  if (RMW_QOS_POLICY_RELIABILITY_UNKNOWN == publisher_qos.reliability) {
+    RMW_SET_ERROR_MSG("Publisher reliability is unknown");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (RMW_QOS_POLICY_RELIABILITY_UNKNOWN == subscription_qos.reliability) {
+    RMW_SET_ERROR_MSG("Subscription reliability is unknown");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (RMW_QOS_POLICY_DURABILITY_UNKNOWN == publisher_qos.durability) {
+    RMW_SET_ERROR_MSG("Publisher durability is unknown");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (RMW_QOS_POLICY_DURABILITY_UNKNOWN == subscription_qos.durability) {
+    RMW_SET_ERROR_MSG("Subscription durability is unknown");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (RMW_QOS_POLICY_LIVELINESS_UNKNOWN == publisher_qos.liveliness) {
+    RMW_SET_ERROR_MSG("Publisher liveliness is unknown");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (RMW_QOS_POLICY_LIVELINESS_UNKNOWN == subscription_qos.liveliness) {
+    RMW_SET_ERROR_MSG("Subscription liveliness is unknown");
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  // Presume profiles are compatible until proven otherwise
+  *compatibility = RMW_QOS_COMPATIBILITY_OK;
+
+  // Initialize reason buffer
+  if (reason) {
+    reason[0] = '\0';
+  }
+
+  // Best effort publisher and reliable subscription
+  if (publisher_qos.reliability == RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT &&
+    subscription_qos.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE)
+  {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    rmw_ret_t append_ret = _append_to_buffer(
+      "ERROR: Best effort publisher and reliable subscription;",
+      reason,
+      reason_size);
+    if (RMW_RET_OK != append_ret) {
+      return append_ret;
+    }
+  }
+
+  // Volatile publisher and transient local subscription
+  if (publisher_qos.durability == RMW_QOS_POLICY_DURABILITY_VOLATILE &&
+    subscription_qos.durability == RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL)
+  {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    rmw_ret_t append_ret = _append_to_buffer(
+      "ERROR: Volatile publisher and transient local subscription;",
+      reason,
+      reason_size);
+    if (RMW_RET_OK != append_ret) {
+      return append_ret;
+    }
+  }
+
+  const rmw_time_t & pub_deadline = publisher_qos.deadline;
+  const rmw_time_t & sub_deadline = subscription_qos.deadline;
+  const rmw_time_t deadline_default = RMW_QOS_DEADLINE_DEFAULT;
+
+  // No deadline for publisher and deadline for subscription
+  if (pub_deadline == deadline_default && sub_deadline != deadline_default) {
```

Comment:
The policies in ROS have this annoying `SYSTEM_DEFAULT` thing which I have always read as the application explicitly "don't care, do what you think is best", which doesn't make much sense to me. Secondly, I've always interpreted that setting as valid only when creating an endpoint, but never as a value returned by the underlying middleware.

If that interpretation is correct (which it may not be, given that the `SYSTEM_DEFAULT` is handled later in this function), then it seems illogical that a default deadline setting would be treated differently here. Doesn't it make far more sense to simply say the default deadline is ∞ (however represented)? That way, there always is a deadline.

If you then define the comparison functions such that ∞ is handled correctly, this all becomes much simpler, IMHO.

Reply:
> Secondly, I've always interpreted that setting as valid only when creating an endpoint, but never as a value returned by the underlying middleware.

Yes, that should be the case.

> If that interpretation is correct (which it may not be, given that the SYSTEM_DEFAULT is handled later in this function), then it seems illogical that a default deadline setting would be treated differently here. Doesn't it make far more sense to simply say the default deadline is ∞ (however represented)? That way, there always is a deadline.

But the "system default" could be different from infinite, e.g. if you specify in a qos profile file a default profile with a different deadline value (which would be weird, but possible).
I would only use this function with "actual" qos profiles , and not with one where an item can be "SYSTEM_DEFAULT", so we can also consider failing if an item is "SYSTEM_DEFAULT".

Reply:
Actually, this new function is failing when a policy has value "UNKNOWN", which is used when an entity has a qos value valid for the underlying middleware that hasn't an equivalent in ROS 2.

I'm not sure why this function should return a compatibility warning when "SYSTEM_DEFAULT" and an error when "UNKNOWN".
I actually think that the opposite makes more sense, or either returning a compatibility warning or error in both.

@jacobperron what do you think?

Reply:
> But the "system default" could be different from infinite, e.g. if you specify in a qos profile file a default profile with a different deadline value (which would be weird, but possible).

I wasn't sure if `RMW_QOS_DEADLINE_DEFAULT` is equivalent to "system default". If it is, then I agree we should treat it the same as other system default values (e.g. warn).

> I would only use this function with "actual" qos profiles , and not with one where an item can be "SYSTEM_DEFAULT", so we can also consider failing if an item is "SYSTEM_DEFAULT".

Unfortunately, all of the [widely used profiles defined in `rmw`](https://github.com/ros2/rmw/blob/ae0553adb574c031343965195ec4a73039e62389/rmw/include/rmw/qos_profiles.h) use "SYSTEM_DEFAULT" values. I think it would be unfortunate if this function failed doing something like this:

```c
rmw_qos_profile_check_compatible( rmw_qos_profile_sensor_data, rmw_qos_profile_sensor_data, ...);
```

Although, producing a warning isn't much better...

Maybe we should change the default values for those profiles :sweat_smile: 

> Actually, this new function is failing when a policy has value "UNKNOWN", which is used when an entity has a qos value valid for the underlying middleware that hasn't an equivalent in ROS 2.

I don't know if this applies here. This is a common DDS implementation that presumes the RMW supports all the ROS 2 QoS settings (we could document it here to be explicit). If an RMW doesn't support all of the QoS policies, then they should probably provide their own implementation of this function, especially if the RMW is not a DDS implementation.

Reply:
What is this function for?

I imagine it being used in two ways:

- Check if the qos profile of two existing endpoints is compatible (for example, using the qos profiles got from the `rmw_get_publishers_info_by_topic`/`rmw_get_subscription_info_by_topic`).
  This use case doesn't seem super useful though, better to use the "Qos incompatible event".
- Check if an user provided profile is compatible with the profile of an existing(s) endpoint (one profile got from the functions above, the other is user provided).
  Useful if you want to "make up" a profile compatible with existing entities.

In case 1, "SYSTEM_DEFAULT" is impossible, and "UNKNOWN" is a possible value.
In case 2, the user provided profile can have a "SYSTEM_DEFAULT" value, but I'm not sure why someone would pass that.

> I don't know if this applies here. This is a common implementation for DDS that presumes the RMW supports all the ROS 2 QoS settings (we could document it here to be explicit). If an RMW doesn't support all of the QoS policies

Yes, and in that case you can still get "UNKNOWN".
e.g.: 
- create a publisher with durability=SYSTEM_DEFAULT
- pass a qos profile file that specifies "DDS_TRANSIENT_DURABILITY_QOS" (using topic filters if desired) (transient is not the same as transient local)
- call `rmw_get_publishers_info_by_topic()`, the qos profile returned will have `durability=UNKNOWN`.

Same applies to any application mixing native DDS with ROS 2.

Reply:
>  Doesn't it make far more sense to simply say the default deadline is ∞ (however represented)? That way, there always is a deadline.

@eboasson I think this is what https://github.com/ros2/rmw/pull/301 is trying to address. Moving away from a "default" value to a more explicit "infinite" value.

AFAICT, passing a "default" value for deadline, liveliness, or lifespan will result in an "infinite" duration (for cyclonedds, fast-dds, and connext). However, this is not what is [documented](https://github.com/ros2/rmw/blob/ae0553adb574c031343965195ec4a73039e62389/rmw/include/rmw/types.h#L427) in `rmw`, hence this special logic. If #301 lands, then I'm inclined to change this implementation as you pointed out.

Reply:
@ivanpauno Yeah, those are the two uses of this function that I had in mind. And the first one is the primary motivation for this API. AFAIK, tools that want to introspect the ROS graph (e.g. ros2doctor and rqt_graph) cannot get "QoS incompatible" events for all nodes in the graph. This API let's them warn about (possible) compatibility issues.

I guess it is technically possible to get a "SYSTEM_DEFAULT" back from `rmw_get_publishers_info_by_topic`. Though it's not very useful, there's nothing that says it can't return such a value. My thought is that this function will report possible compatibility issues to the best of it's ability. If it gets a policy value of "SYSTEM_DEFAULT" or "UNKNOWN" then the best we can do is warn about possible incompatibilities. Originally, I didn't think that it was possible to get "UNKNOWN" values from `rmw_get_publishers_info_by_topic` / `rmw_get_subscription_info_by_topic`, but if that is the case then I agree we should change the behavior of this function to produce a warning instead.

Reply:
> AFAIK, tools that want to introspect the ROS graph (e.g. ros2doctor and rqt_graph) cannot get "QoS incompatible" events for all nodes in the graph

Ah yes, that's true.

> I guess it is technically possible to get a "SYSTEM_DEFAULT" back from rmw_get_publishers_info_by_topic

`rmw_get_publishers_info_by_topic` should not return "SYSTEM_DEFAULT", but a more specific value.

> but if that is the case then I agree we should change the behavior of this function to produce a warning instead.

Using a warning for both system default and unknown sounds good.

Reply:
Warning on "unknown" values: 6c789b4

Reply:
> > I guess it is technically possible to get a "SYSTEM_DEFAULT" back from rmw_get_publishers_info_by_topic

> rmw_get_publishers_info_by_topic should not return "SYSTEM_DEFAULT", but a more specific value.

I think that `rmw_get_publishers_info_by_topic` should try and avoid returning system default, or shouldn't ever. The idea is that system default is a stand-in for "some externally specified setting", so when querying it for a real entity, it should just return the ROS equivalent or unknown if there isn't an equivalent.

Reply:
To circle back to the original comment, I think I'll leave the current implementation as-is since it seems to agree with the RMW API docs and https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/#qos-compatibilities.
However, I do think that replacing the default duration value with an explicit infinity value would be better (related to https://github.com/ros2/rmw/pull/301). Pending further changes to the RMW API, I'm happy to come back and simplify the logic in this implementation.

While endpoints may not return "SYSTEM_DEFAULT" values, it's possible that other use-cases not involving endpoint queries will call this function. So I think it's nice to warn about "SYSTEM_DEFAULT"  values, instead of returning an error with no info about other possible incompatibilities.